docs/source/class_source.rst

```diff
@@ -0,0 +1,335 @@
+
+An SNR based approach is also provided, and a Bandpass method should
+be implemented in a future release.  Again, these components are fully
+customizable and the base interfaces were designed with algorithm
+developers in mind.
+
+Implementing concrete ``GlobalClassSelector`` leverage subcomponents
+described below.
+
+.. mermaid::
+
+   classDiagram
+       ClassSelector <|-- GlobalClassSelector
+       class GlobalClassSelector{
+	   averager: Averager2D
+	   function: ImageQuaityFunction
+	   heap_size: int
+	   }
+       GlobalClassSelector *-- ImageQualityFunction
+       GlobalClassSelector ..> Heap
+       GlobalClassSelector <|-- GlobalWithRepulsionClassSelector
+
+
+       class ImageQualityFunction{
+	  -_function
+	  +__call__()
+	  }
+       ImageQualityFunction o-- WeightedImageQualityMixin
+       ImageQualityFunction <|-- BandedSNRImageQualityFunction
+       ImageQualityFunction <|-- ContrastImageQualityFunction
+       ImageQualityFunction <|-- BandpassImageQualityFunction_TBD
+
+       class WeightedImageQualityMixin{
+	   -_weight_function
+       }
+       WeightedImageQualityMixin <|-- RampWeightedImageQualityMixin
+       WeightedImageQualityMixin <|-- BumpWeightedImageQualityMixin
+
+       GlobalClassSelector <|-- RampWeightedContrastImageQualityFunction
+       RampWeightedImageQualityMixin <|-- RampWeightedContrastImageQualityFunction
+       GlobalClassSelector <|-- BumpWeightedContrastImageQualityFunction
+       BumpWeightedImageQualityMixin <|-- BumpWeightedContrastImageQualityFunction
+
+Class Repulsion
+^^^^^^^^^^^^^^^
+
+Class Repulsion are techniques used to avoid classes based on some
+criterion.  Currently we provide ``GreedyClassRepulsion``, but this
+mix-in class can be mimicked to implement alternate schemes.
+
+``GreedyClassRepulsion`` is based on the following intuition. Assume
+the selection has in fact ordered the classes s.t. *the "best"
+classes occur first*. It follows that the "best" expression of a
+viewing angle locus will be the first seen.  Now assume *the
+classifier returns classes with closest viewing angles* (up to
+reflections).  Then the classes formed by *neighbors of the current
+expression are inferior*.  The aggressiveness of the neighbor
+repulsion count is tunable.
+
+In practice, ``GreedyClassRepulsion`` is a mix-in designed to be mixed
+into any other ``ClassSelector``.  Note, that repulsion can (and will)
+dramatically reduce the population of class averages returned.
+
+
+Image Quality Functions
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``ImageQualityFunction`` interface provides a consistent way to
+bring your own function to measure the quality of an aligned and
+registered class average.  This function should operate on a single
+Image, with conversions and broadcasting being handled behind the
+scenes.
+
+An example would be ``ContrastImageQualityFunction`` which computes
+and returns contrast as variance.
+
+Another advantage of using the class is that it exposes and manages a
```

Comment:
Not sure what you mean here?

Reply:
I implemented it (global grid cache), but there is a hang up implementing it using the native python builtins when parallel processing. I'm waiting for that to be resolved by upstream packages.  I think maybe resolved py3.9 forward.  Because it is being fixed upstream, I'm not sure we should spend time on it.

Reply:
(a temp work around is also one of the projects we advertised towards interns)